### ISSUE TITLE
[learning][research] Define differentiability boundaries and add surrogate-guided local pose refinement

### ISSUE THREAD

**Issue by JanDuchscherer104** [JanDuchscherer104/ARIA-NBV#77]
Parent: #67

Depends on: #74 for the hard-oracle environment, #75 for a learned finite-candidate value model, and #76 for bounded residual proposals.

## Outcome

Make the differentiability contract explicit and implement the **highest-ROI differentiable path**: gradient-based local refinement through a learned actor-visible RRI/`Q_H` surrogate, followed by hard feasibility and hard-oracle validation.

Do not attempt to backpropagate through the full current renderer → hard mask → ragged point cloud → crop → voxel fusion → nearest-surface RRI chain as the first solution. Direct soft-oracle differentiation remains an optional, separately named research ablation.

### A. Offline one-step or finite-horizon value learning

Only the learned scorer/value model must be differentiable with respect to its parameters:

```math
\nabla_\theta Q_{H,\theta}(s,e,q_i).
```

Candidate sampling, hard validity, rendering, target cropping, fusion, RRI, action selection, and bootstrap argmax may all be non-differentiable data/environment operations.

### B. Discrete stochastic selection with policy gradients

Only the selected action log probability must retain gradients:

```math
\nabla_\theta J
=\mathbb E[(G-b)\nabla_\theta\log\pi_\theta(a\mid s,Q,m)].
```

The oracle reward can be an opaque scalar from the hard environment. No renderer gradient is required.

### C. Stochastic continuous proposal learning

Two valid estimators exist:

1. **score-function gradient:** differentiate `log p_phi(q|s,e)`; the oracle remains black-box;
2. **pathwise/surrogate gradient:** reparameterize `q=f_phi(epsilon,s,e)` and differentiate a learned surrogate with respect to `q`.

The first is unbiased but can have high variance. The second is lower variance but optimizes the surrogate and therefore requires hard-oracle validation.

### D. Gradient-based local pose refinement

This requires a differentiable pose parameterization and a surrogate value field:

```math
q' = \operatorname{Project}_{C}
\left(q + \eta\nabla_q\widehat Q_\psi(s,e,q)\right).
```

The final projection, hard mask, and hard-oracle evaluation may remain non-differentiable.

### E. Direct gradients through the true RRI oracle

Then essentially the entire chain must carry stable pose gradients:

```text
pose
→ camera/raster visibility
→ depth
→ unprojection
→ target membership
→ point fusion
→ point–surface distance
→ RRI
```

The current implementation does not satisfy this in a useful smooth sense.

## Why the current hard oracle is not an appropriate gradient path

The present pipeline includes:

- hard z-buffer rasterization with one closest face;
- discrete `pix_to_face` hit masks and near/far comparisons;
- boolean valid-pixel indexing and ragged point counts;
- hard OBB/AABB membership tests and face selection;
- voxel `floor`, `unique`, deterministic point capping, and discrete lengths;
- hard nearest-face and nearest-point assignments in point–mesh distance;
- hard candidate validity and argmax/selection.

Some subexpressions are piecewise differentiable, but visibility, crop, voxel, and nearest-neighbor boundaries produce discontinuities or sparse/unstable gradients.

Stochastic computation graphs formalize cases where the “loss function is defined by an expectation over a collection of random variables”; they support score-function estimators without differentiating the external world.

Soft Rasterizer notes that standard rendering contains “a fundamental discretization step called rasterization” and replaces hard triangle assignment with probabilistic aggregation. Such a renderer would define a different soft oracle, not retroactively make the current RRI label differentiable.

## Required explicit modes

Add a typed experiment setting:

```python
class ProposalGradientMode(StrEnum):
    BLACK_BOX_ORACLE = "black_box_oracle"
    SCORE_FUNCTION = "score_function"
    SURROGATE_PATHWISE = "surrogate_pathwise"
    SOFT_ORACLE_ABLATION = "soft_oracle_ablation"
```

Every collector/trainer must state its mode in manifests and checkpoints. No mode may silently fall back to another.

## Default gradient boundary

For `BLACK_BOX_ORACLE` and `SCORE_FUNCTION`:

```python
with torch.no_grad():
    oracle_result = env.evaluate(...)
```

or explicitly detach all oracle results at the environment boundary. Preserve live policy/proposal log probabilities separately.

Add gradient-leak tests proving that GT mesh, rendered depth, oracle points, target crops, and RRI tensors have no path into actor parameters.

### Surrogate target

Train an actor-visible model to approximate immediate target gain or finite-horizon value:

```math
\widehat Q_\psi(s,e,q)
\approx Q_H(s,e,q).
```

Use hard-oracle labels from valid candidates. The surrogate may share the selector architecture from #75 but must expose gradients with respect to a continuous local pose descriptor.

### Pose parameterization

Optimize a local 5-DoF tangent/residual:

```text
(delta_x, delta_y, delta_z, delta_yaw, delta_pitch)
```

in a declared root-, target-, or anchor-local frame. Roll remains fixed for the canonical action space.

Avoid optimizing raw 12D pose matrices. Use continuous rotation construction and re-orthogonalization through the existing pose utilities.

### Refinement algorithm

For each promising anchor:

1. initialize `delta=0` or the proposer residual;
2. evaluate surrogate value and actor-side soft feasibility;
3. take 1–`K` bounded gradient steps;
4. project parameters into family/motion bounds;
5. construct the physical pose;
6. apply hard geometric feasibility;
7. deduplicate and place into the final reservoir/shell;
8. evaluate with the canonical hard oracle for labels/reporting.

Example objective:

```math
L_{refine}(\delta)=
-\widehat Q_\psi(s,e,q(\delta))
+\lambda_{step}\|\delta\|_2^2
+\lambda_{feas}\,\phi_{soft-invalid}(q(\delta))
+\lambda_{dup}\,\phi_{duplicate}(q(\delta)).
```

The soft feasibility term guides optimization but cannot override the hard mask.

### Trust region

Refinement must remain local:

```text
max translation residual
max yaw/pitch residual
max gradient steps
step-size schedule
minimum predicted improvement
fallback to original anchor
```

Persist every value and the pre/post pose.

## Surrogate validation

Before allowing pose gradients to influence proposals, validate:

- held-out value/ranking error;
- calibration versus pose perturbation size;
- gradient finite-difference agreement;
- gradient direction agreement with small hard-oracle perturbation probes;
- out-of-distribution detection for unsupported scene/target/candidate regions;
- improvement rate after hard feasibility and hard-oracle rescoring;
- failure rate where surrogate ascent reduces true RRI.

A surrogate with good candidate ranking may still have bad local pose gradients. Gradient validation is a separate gate.

## Optional soft-oracle ablation

Only after surrogate refinement is established, consider a separately named differentiable approximation using:

```text
soft/probabilistic rasterization or differentiable ray marching
fixed-size weighted ray samples rather than ragged hard hits
soft target-OBB membership weights
soft voxel splatting or continuous occupancy/SDF state
softmin/log-sum-exp point–surface distances
smooth feasibility penalties
```

Required rules:

- the soft score has a new metric/provenance name;
- it never overwrites canonical hard target RRI;
- its bias and gradient agreement are measured against hard-oracle perturbations;
- final candidate and policy evaluation uses the hard oracle;
- source code and settings are versioned independently.

## API additions

```python
class DifferentiableCandidateSurrogate(nn.Module):
    def forward(
        self,
        actor_state: ActorState,
        target: ActorTargetDescriptor,
        candidate_pose_local_5d: Tensor,
    ) -> SurrogatePrediction: ...

class LocalPoseRefiner:
    def refine(
        self,
        anchors: CandidateAnchorBatch,
        actor_state: ActorState,
        target: ActorTargetDescriptor,
    ) -> RefinedCandidateBatch: ...
```

`RefinedCandidateBatch` carries:

```text
anchor identity
pre/post local parameters and poses
surrogate values before/after
gradient norms and step count
trust-region/projected flags
soft-feasibility values
hard-validity result
hard-oracle value when queried
surrogate and refiner checkpoint hashes
```

## Acceptance criteria

- [ ] The four gradient modes are explicit and manifest-bound.
- [ ] Default offline/online finite-candidate learning detaches the oracle boundary.
- [ ] Policy-gradient collection preserves live selected log probabilities without oracle gradients.
- [ ] A differentiable actor-visible surrogate exposes finite gradients with respect to local 5-DoF pose.
- [ ] A bounded local refiner improves surrogate values while respecting trust-region and frame contracts.
- [ ] Every refined pose passes the ordinary hard feasibility/final-shell pipeline.
- [ ] Surrogate gradients pass finite-difference and hard-oracle perturbation diagnostics.
- [ ] Promotion requires improvement under the hard proposal benchmark and endpoint oracle.
- [ ] Any soft-oracle implementation has a separate metric/contract and cannot replace hard RRI.
- [ ] Actor/oracle leakage tests prove no GT render/crop field reaches inference.

## Required tests

- hard-oracle `requires_grad=False` boundary;
- score-function update with a deliberately non-differentiable reward fixture;
- local pose transform/rotation continuity;
- surrogate pose finite differences;
- zero-step/refinement identity;
- trust-region projection and gradient clipping;
- hard-invalid refined candidate rejection;
- original-anchor fallback;
- duplicate handling after refinement;
- checkpoint/contract provenance round trip;
- soft-oracle metric-name and reader separation;
- hard-oracle post-refinement evaluation.

## Evidence before promotion

Report:

- surrogate ranking/calibration;
- finite-difference gradient error;
- hard-oracle directional agreement under small perturbations;
- percentage of anchors improved/degraded by hard RRI;
- improvement magnitude and proposal regret;
- invalidity and duplicate rates after refinement;
- runtime per gradient step and candidate;
- sensitivity to trust region, steps, and optimizer;
- OOD/failure strata;
- hard versus optional soft-oracle score/gradient bias.

## External sources

- Schulman et al., *Gradient Estimation Using Stochastic Computation Graphs*: https://arxiv.org/abs/1506.05254
- Liu et al., *Soft Rasterizer*: https://openaccess.thecvf.com/content_ICCV_2019/html/Liu_Soft_Rasterizer_A_Differentiable_Renderer_for_Image-Based_3D_Reasoning_ICCV_2019_paper.html
- Ravi et al., *PyTorch3D*: https://arxiv.org/abs/2007.08501

## Non-goals

- Direct end-to-end differentiation through the current hard RRI pipeline is not required for `Q_H`, selector learning, AWR, or REINFORCE.
- The soft oracle is not the canonical label or endpoint metric.
- Gradient-based refinement does not bypass hard feasibility.
- This issue does not turn online continuous control into a thesis-core claim.

**Comment by JanDuchscherer104** [comment]
## Research enrichment — round 2: nonsmooth controls, OOD-aware gradients, and perturbation-budget evaluation

### 1. Require derivative-free controls for every gradient-based claim

The hard oracle is piecewise/nonsmooth because visibility, hit masks, crop membership, voxelization, nearest surfaces, and hard feasibility change discretely. Even a smooth surrogate can have misleading gradients outside its training support.

Compare local surrogate ascent against equal-query/equal-start controls:

```text
random local perturbation
finite coordinate search
Cross-Entropy Method
CMA-ES or another named derivative-free optimizer if justified
```

CEM references:

- https://doi.org/10.1023/A:1010091220143
- https://doi.org/10.1007/s10479-005-5724-z

A gradient method is useful only if it improves hard-oracle proposal quality beyond these controls under equal proposal/evaluation budget.

### 2. Validate gradients locally, not only predictions globally

A surrogate can rank sampled candidates well yet provide poor off-grid gradients. Add three separate gates:

1. **autograd versus finite difference** on the surrogate;
2. **surrogate gradient direction versus hard-oracle directional probes** around valid anchors;
3. **realized hard-oracle improvement after projected steps**.

For each anchor sample small positive/negative perturbations in every local 5-DoF axis and report:

```text
cosine/directional agreement
sign agreement
finite-difference curvature
fraction crossing hard visibility/validity boundaries
true improvement/degradation rate
```

### 3. Add uncertainty- and support-aware trust regions

Do not refine when the surrogate is extrapolating. Use an ensemble, evidential head, nearest-training-support score, or another calibrated OOD diagnostic. Resolve trust region and step count from support:

```text
high support/low uncertainty → bounded refinement
low support/high uncertainty → smaller step or original-anchor fallback
outside support → no refinement
```

Persist the support and fallback decision. Hard-oracle results must be stratified by surrogate uncertainty/OOD.

### 4. Stop gradients at hard projection and final-shell selection

The canonical path should be:

```text
continuous local parameters
→ differentiable pose constructor
→ surrogate objective
→ optimizer step
→ bounded parameter projection (stop-gradient semantics declared)
→ hard feasibility
→ dedup/final-shell selection
→ hard oracle
```

Do not silently use a straight-through estimator through clipping, masks, collision decisions, or discrete subset selection. Any STE must be a separately named ablation with bias acknowledged.

### 5. Use source-separated soft feasibility

Guthrie's differentiable signed-distance formulation is relevant for continuous collision penalties:

https://arxiv.org/abs/2302.09704

However, actor-side refinement cannot query a GT-mesh SDF at inference. Keep:

```text
oracle_gt_sdf penalty — privileged refinement upper bound
actor_belief_sdf / occupancy penalty — actor-admissible
learned feasibility penalty — actor-admissible if calibrated
```

Every result states which one guided refinement. Final hard GT feasibility remains offline evaluation only.

### 6. Bound optimization by a complete perturbation budget

Report not only gradient steps but total work:

```text
number of anchors
surrogate queries
backward passes
hard feasibility queries
hard oracle queries
wall time and peak memory
```

Compare methods under matched budgets and starts. A method using 20 anchors × 10 gradient steps is not directly comparable to one 60-row static shell without accounting for compute.

### 7. Evaluate local improvement and final set improvement separately

Per-anchor ascent can create duplicate candidates or collapse the set. Report:

```text
pre/post surrogate and hard value per anchor
hard-validity survival
set duplicate rate
best-of-K / proposal regret after final-shell selection
target-direction/family diversity
```

The final scientific object is the realized candidate set and endpoint policy, not the average surrogate increase.

### 8. Treat a soft renderer/oracle as a distinct biased objective

Soft Rasterizer explains that ordinary rasterization contains a discrete assignment and replaces it with probabilistic aggregation:

- https://openaccess.thecvf.com/content_ICCV_2019/html/Liu_Soft_Rasterizer_A_Differentiable_Renderer_for_Image-Based_3D_Reasoning_ICCV_2019_paper.html

A soft visibility/crop/distance pipeline changes the objective. It requires:

```text
new metric name and contract hash
bias against canonical hard RRI
hard-oracle directional agreement
temperature/smoothing sensitivity
failure at occlusion/crop boundaries
```

It may guide proposals, but never overwrite canonical labels or endpoint evaluation.

### 9. Add pathological gradient fixtures

Include:

```text
candidate just before/after an occlusion boundary
near-frustum clipping boundary
candidate crossing target OBB crop boundary
nearest-face switch
voxel-boundary crossing
hard feasibility boundary
near-solved root normalization
```

These are precisely where visually plausible gradients can become unstable or semantically wrong.

### Additional acceptance criteria

- [ ] Every surrogate-gradient result has equal-budget derivative-free controls.
- [ ] Surrogate autograd, finite differences, and hard-oracle directional probes are distinct gates.
- [ ] OOD/uncertainty controls trust-region size and anchor fallback.
- [ ] No undeclared straight-through gradient crosses hard projection, masks, or subset selection.
- [ ] Actor and privileged soft-feasibility sources are separated.
- [ ] Optimization reports complete query/compute budgets.
- [ ] Final candidate-set proposal regret and diversity are measured after refinement.
- [ ] A soft oracle has a separate biased metric contract and hard-oracle validation.
- [ ] Visibility, crop, voxel, nearest-surface, and validity boundary fixtures are required.
